### ISSUE TITLE
RemoteTech enhancement request, 

### ISSUE THREAD

**Issue by erendrake** [KSP-KOS/KOS#744]
This is a conversation from IRC, i am recording it here while its still in my brain. Some of it is a dupe of another issue.

<LabMonkey> Erendrake: tbh, (and I'm not this far along yet, but) I'm considering how to print/echo user input in the local display and still allow the server to do the echoing without doubling the output in the local display.  In my particular case, it may be useful to do this since if I do something like changing text or background color depending on whether the input has been echoed then that can serve as a visual queue to the user
<LabMonkey> as to the state of command processing with delay taken into consideration.
* medsouz|offline is now known as medsouz
<LabMonkey> Erendrake: on a related note... Some of the information I've read via google-found wikis and the like suggest that RT2 delay is not taken into consideration by kOS.  Is that true?
<LabMonkey> Erendrake: What I think would be an ideal scenario is that if a probe is out some few light seconds or minutes from the nearest control point, my console will allow the user to see what they're typing and even issue the commands, but they'll be highlighted to indicate that they have not been acknowledged/echoed by the server until that echo is received, at which point the command will no longer be highlighted.
<LabMonkey> Erendrake: So it's not just network latency between kOSKontrol and kOS, but latency between the controlling pod/ksc and the probe

**Comment by ascensionislander** [comment]
Were such a thing implemented, would we still see the craft in question execute the commands given to it after the one-way delay time? I think it'd be weird to see things as though the camera had instant communication with KSC, but the terminal window had to transmit back.

Then again, that'd make it pretty cool if you could run KSP on one monitor or computer and a remote telnet on another, and then position yourself so you could only see the monitor or computer with the telnet on it. <crazy proposal> And then you could have a command to take an image from, say, a Hullcam VDS camera part and transmit it back after the one-way delay</crazy proposal>. Of course, this would be super-extreme-hardcore mode.

**Comment by abenkovskii** [comment]
@ascensionislander Not so crazy proposal: camera grabbing should be in a separate mod that has it's own telnet interface and it's own RT integration. The kOS is an autopilot, not a video capturing software. What you are talking about isn't easy to implement and only a couple of kOS users would actually use this feature. A separate mod for "spectating" with optional RT delay, integration with all existing camera mods and a bunch of other features will probably attract more users.
P.S. Another variant is to extend Hullcam VDS to add RT integration and telnet.

**Comment by labmonkey42** [comment]
I would rather it be that whatever camera plugin implement optional RT2 delay, and if you're going for superhardcoremlgproextremeeditionop2015omgwtfbbq immersion then you can run KSP with Probe Control Room and just add a RPM screen for the kOS console.

In the case of the text Erendrake copied above, the goal is to create an external program which will automate the automation over telnet (e.g. copy program from 0 to specific CPU, parse flight data from program output and issue generated commands to update flight parameters, switch programs when changing flight phase, etc).  It's all rather SuperSekrit(TM) right now, but in order for it to work at all (and more so properly) I need to be sure I understand how the kOS telnet server works.

Also... Thanks Erendrake!

**Comment by Dunbaratu** [comment]
I am dead set against implementing radio delay in such a fashion that it only exists in the telnet terminal but not in the onscreen terminal.  Those two are meant to be identical and you choose which you wanted based on outside-the-game considerations like how to arrange your gaming desktop. The network connection between KSP and your telnet client is emphatically *not* an  in-gameworld phenomenon.  

Any radio delay needs to be implemented outside the telnet code, in a way that is happening to the gui terminal too.

**Comment by erendrake** [comment]
I agree with @Dunbaratu on the delay terminal issue. the in game terminal IS telnet terminal, just in game. Both terminals should, whenever possible, behave exactly the same.

im going to think about the roundtrip delay idea. i think its a good one :)

**Comment by Dunbaratu** [comment]
I like the delay (again, for both terminals equally), but it has a few connected consequences to go with it we have to think over carefully:

1 - In exactly the same way that remote-tech delay shouldn't affect what a meat pilot sitting in the cockpit can do, it also shouldn't prevent the meat pilot from typing into the computer console quickly without delay.  The delay should be a matter of finding the delay of the signal path to the nearest *kerbal* who can operate the console, who is not necessarily all the way back at KSC.  (i.e. if there is a kerbal in the cockpit - no delay.  If there is a kerbal up in the command capsule in orbit of Duna, talking to a satellite of Duna a few million miles above duna, maybe a second or two delay.. If there is no Kerbal around at all, full delay back to KSC).

2- Maybe this could be something fun to have differentiate between kerbal's jobs.  A *pilot* can use the console as-is and type commands. But it takes an *engineer* to actually edit the code locally without talking to the KSC - anyway just a thought.  So you'd be finding the delay to the nearest Kerbal who can do the job (assuming all 3 jobs existing back at KSC).

**Comment by abenkovskii** [comment]
1. scientists are still obsolete, engineers are most useful (in KSP)
2. Newer tested it but it is documented [here](http://remotetechnologiesgroup.github.io/RemoteTech/#command-stations) that a player can set up an off-Kerbin command center and than radio delay would be calculated from the nearest command station.
3. RT radio delay for ships with kerbals in command seats is always 0.
4. How those two things are exposed through API?
